Add radix sort

LilithHafner opened the pull request:
I've implemented radix sort and integrated it into the Julia sorting architecture. Micro-benchmarking reports a substantial speedup in many cases and minor regressions in a few cases. I hope that with a bit of collaboration, we can minimize those regressions as well.

This grows out of a [discussion on discourse](https://discourse.julialang.org/t/request-for-comments-a-roadmap-to-faster-sort-for-primitive-types-using-radixsort/70269) and a [PR to SortingAlgorithm.jl](https://github.com/JuliaCollections/SortingAlgorithms.jl/pull/53). It is inspired in part by the radix sort algorithm in SortingAlgorithms.jl. This is different from the radix sort algorithm in SortingAlgorithms.jl because it is substantially faster in most cases and more easily extensible.

This PR is based on a fast radix sort for unsigned integers in increasing order. To sort other types and orders—currently all signed and unsigned integers, Float32, Float64, and Char, for Forward and Reverse orderings are supported—we have a function `Sort.Serial.serialize` that converts objects to unsigned integers which are sorted and then the original objects are reconstructed via `Sort.Serial.deserialize`. These two functions come with a trait `Sort.Serial.Serializable` that indicates whether a type is serializable for a given ordering. If a package defines a custom type that can be sorted via radix sort, it may either a) do nothing and we will continue to dispatch to a comparison sort at compile time or b) define these three functions and we will dispatch to radix sort when appropriate.

I have left out many features and minor optimizations that have come up in discussion surrounding this algorithm, preferring simplicity over minor performance benefits. Those optimizations can be added now, later, or never.

A small sample of benchmarking vs. mar 1 master (3a5dc09ce4)
```
@benchmark sort!(x) setup=(x=rand(1:1, 1)) evals=1 # 50%, 48% speedup
@benchmark sort!(x) setup=(x=rand(1:30, 30)) evals=1 # 36%, 28% speedup
@benchmark sort!(x) setup=(x=rand(1:1000, 1000)) evals=1 # 479%, 450% speedup
@benchmark sort!(x) setup=(x=rand(1:30000, 30000)) evals=1 # 635%, 793% speedup
@benchmark sort!(x) setup=(x=rand(1:1000000, 1000000)) evals=1 # 648%, 642% speedup

@benchmark sort!(x) setup=(x=rand(1)) evals=1 # 71%, 17% speedup
@benchmark sort!(x) setup=(x=rand(10)) evals=1 # 6%, 9% speedup
@benchmark sort!(x) setup=(x=rand(30)) evals=1 # 54%, 14% speedup
@benchmark sort!(x) setup=(x=rand(1000)) evals=1 # 76%, 73% speedup
@benchmark sort!(x) setup=(x=rand(30000)) evals=1 # 121%, 119% speedup
@benchmark sort!(x) setup=(x=rand(1000000)) evals=1 # 139%, 144% speedup

@benchmark sort!(x) setup=(x=rand(Char, 1)) evals=1 # 75%, 155% speedup
@benchmark sort!(x) setup=(x=rand(Char, 30)) evals=1 # 32%, 86% speedup
@benchmark sort!(x) setup=(x=rand(Char, 1000)) evals=1 # 272%, 284% speedup
@benchmark sort!(x) setup=(x=rand(Char, 30000)) evals=1 # 771%, 873% speedup
@benchmark sort!(x) setup=(x=rand(Char, 1000000)) evals=1 # 1002%, 965% speedup


@benchmark sort!(x) setup=(x=rand(Bool, 10^6)) evals=1 # 1843%, 2123% speedup
@benchmark sort!(x) setup=(x=rand(Int8, 10^6)) evals=1 # 5368%, 5272% speedup
@benchmark sort!(x) setup=(x=rand(UInt8, 10^6)) evals=1 # 5275%, 5162% speedup

@benchmark sort!(x) setup=(x=rand(Int16, 10^6)) evals=1 # 2204%, 2165% speedup
@benchmark sort!(x) setup=(x=rand(UInt16, 10^6)) evals=1 # 2197%, 2140% speedup
@benchmark sort!(x) setup=(x=rand(Float16, 10^6)) evals=1 # 2%, -1% speedup

@benchmark sort!(x) setup=(x=rand(Int32, 10^6)) evals=1 # 567%, 548% speedup
@benchmark sort!(x) setup=(x=rand(UInt32, 10^6)) evals=1 # 524%, 525% speedup
@benchmark sort!(x) setup=(x=rand(Char, 10^6)) evals=1 # 1001%, 944% speedup
@benchmark sort!(x) setup=(x=rand(Float32, 10^6)) evals=1 # 491%, 464% speedup

@benchmark sort!(x) setup=(x=rand(Int64, 10^6)) evals=1 # 120%, 118% speedup
@benchmark sort!(x) setup=(x=rand(UInt64, 10^6)) evals=1 # 104%, 106% speedup
@benchmark sort!(x) setup=(x=rand(Float64, 10^6)) evals=1 # 141%, 142% speedup

@benchmark sort!(x) setup=(x=rand(Int128, 10^6)) evals=1 # 12%, 6% speedup
@benchmark sort!(x) setup=(x=rand(UInt128, 10^6)) evals=1 # 1%, -2% speedup
```
(run 2 times, speedups computed by dividing minimum time on master by minimum time on ffcc043)

Problems & todos

- [x] `@benchmark sort!(x) setup=(x=rand(1000))` is 300ns slower for radix sort than quick sort.
It should immediately dispatch to quick sort based only on length, so it is unclear why there 
is such a substantial slowdown. This is probably related to the 100% slowdown for sorting 
single floats. (addressed in https://github.com/JuliaLang/julia/pull/44230#issuecomment-1048455897)

- [x] I've run millions of tests with SortMark.jl, but this still needs efficient comprehensive 
testing in Base, and more hand crafted tests may still find bugs.

- [x] Sort.Serial is a bit of a name conflict with the existing package Serialization
- [ ] Perhaps we should support passing in the temporary array. Is that performance feature used anywhere?

- [x] The heuristics right now are pretty primitive. Perhaps they can be improved or tuned.
- [ ] Potentially rework integration with `fpsort!`

- [ ] Perhaps we should integrate this with sortperm, right now it is incompatible, but with some changes, it could work.

LilithHafner committed:
Add radix sort

oscardssmith commented on base/sort.jl:
```diff
@@ -652,6 +669,85 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::PartialQuickSort,
+    Serial.Serializable(eltype(v), o) === nothing && return sort!(v, lo, hi, a.fallback, o)
+    hi <= lo && return v
+    hi-lo <= SMALL_THRESHOLD && return sort!(v, lo, hi, SMALL_ALGORITHM, o)
+    (hi - lo < 2000 || (hi - lo < 200_000 && sizeof(eltype(v)) == 16)) && return sort!(v, lo, hi, a.fallback, o)
```
isn't `200_000` a really high fallback?

quinnj commented:
Yeah, this is exciting. I'll link to my own implementation that lives in the [InlineStrings.jl](https://github.com/JuliaStrings/InlineStrings.jl/blob/2a0a33cccdba6338585e4f11e990e42267f5334c/src/InlineStrings.jl#L720) package.

Some notable changes I made to the SortingAlgorithms.jl implementation:
  * It operates on any "primitive" type, so no specific restrictions on integer vs float vs anything else
  * I did some extensive benchmarking around the # of bits of the element type and at what point radix sort was suboptimal, findings [here](https://github.com/JuliaStrings/InlineStrings.jl/blob/2a0a33cccdba6338585e4f11e990e42267f5334c/src/InlineStrings.jl#L688)
  * I split out a [`Radix`](https://github.com/JuliaStrings/InlineStrings.jl/blob/2a0a33cccdba6338585e4f11e990e42267f5334c/src/InlineStrings.jl#L700) struct that makes it easy to play around with different radix sizes and stuff, which was helpful in benchmarking
  * Lots of code comments to explain what everything does! It's a trickier algorithm w/ lots of indirection, so I found it helpful to keep track of what's going on.
  
  For InlineStrings.jl, there are fixed-size string types from 1-byte all the way up to 255-bytes, so there were a lot of permutations of performance to check; here's one sample graph of the `String7` type.
  
  Anyway, excited to see the progress in this direction!
![image](https://user-images.githubusercontent.com/2896623/154582419-400f5371-6eec-4575-b3de-eede6efb76b5.png)

LilithHafner commented:
@quinnj's thresholds (64 bits -> 500 & 128 bits -> 2 million) line up okay with the thresholds I've chosen. I expect that the thresholds and heuristics in this PR could be tuned much more preciesly. If you would like to play around with thresholds, everything runs as is with Revise, I think it's just not building from source with the PR.

@quinnj, are there any changes you would recommend here based on your experience with InlineStrings?

LilithHafner committed:
Try to fix build error (1)

LilithHafner committed:
Try to fix build error (2)

LilithHafner committed:
Try to fix build error (3)

LilithHafner committed:
Try to fix build error (4)

LilithHafner committed:
Try to fix build error (5)

LilithHafner committed:
Try to fix build error (6)

LilithHafner committed:
@LilithHafner Try to fix build error (7)

LilithHafner committed:
Try to fix build error (8)

LilithHafner committed:
Remove set default for debugging

LilithHafner committed:
Update sort_int_range! tests

LilithHafner committed:
put back update defalg after bootstrapping (ugly)

LilithHafner committed:
update usage of sort_int_range! in test to add lo, hi

LilithHafner committed:
whitespace

LilithHafner committed:
add RadixSort tests

LilithHafner commented:
Hooray! Everything but whitespace passed

LilithHafner committed:
more radix sort tests

LilithHafner committed:
fix typo

LilithHafner committed:
maintain backwards compatability in sort_int_range

LilithHafner committed:
fix maybereverse use

LilithHafner committed:
remove set defalg

LilithHafner commented:
I'm giving up on understanding the ci errors and switching to blind bisection. I'm going to start by rebasing to the latest master and reverting everything.

LilithHafner commented:
@DilumAluthge & @JuliaLang/github-actions I made a GitHub mistake and accidentally requested a review from you. Feel free to ignore. (or feel free to review if you'd like!)

LilithHafner committed:
Merge branch 'JuliaLang:master' into patch-9

LilithHafner commented:
Woo hoo!!! All checks passed. I was getting weird seemingly inconsistent errors on various systems and test configurations, but rebasing onto the latest master fixed them.

LilithHafner committed:
put back set defalg

quinnj commented:
> @quinnj, are there any changes you would recommend here based on your experience with InlineStrings?

I think the main change I liked in my implementation was making it generic to any `primitive type` types, instead of the back and forth dance w/ `UInt` and having to define a converter interface function. In my mind, radix sort is all about fancy bit tricks, so it makes sense to have it "just work" on any "bits" type ("bits" here referring to `primitive` not `isbitstype`). While performance tuning, I also found it produced drastically simpler code since the algorithm just uses LLVM bit-size primitives.

LilithHafner commented:
@quinnj, How did you get it to put floats in the right order? I'm not sure which LLVM primitives allow radixing into floating point numbers. For reference:
```
julia> bitstring.(-2.0:2.0)
5-element Vector{String}:
 "1100000000000000000000000000000000000000000000000000000000000000"
 "1011111111110000000000000000000000000000000000000000000000000000"
 "0000000000000000000000000000000000000000000000000000000000000000"
 "0011111111110000000000000000000000000000000000000000000000000000"
 "0100000000000000000000000000000000000000000000000000000000000000"
```
What is your preferred way of dealing with primitive types that don't have the same sort order as their binary representations? (Typically 10–30% of runtime is spent converting to/from UInts and 70-90% is spent in UInt forward radix sorting. Even 10% savings is substantial in this domain.)

LilithHafner commented:
I reviewed the implementation in InlineStrings.jl and ran into this issue https://github.com/JuliaStrings/InlineStrings.jl/issues/24 which is a result of making sorting generic to any type.

LilithHafner committed:
dispatch to InsertionSort _before_ fpsort!ing

LilithHafner commented:
On an unrelated note, I've reduced the performance problem on small inputs to the following MWE:
```
f(x) = x == 0 ? 0 : x
g(x) = x == 0 ? 0 : binomial(2x, x)
@btime f(x) setup=(x=$(Ref(0))[]) # 1.514 ns
@btime g(x) setup=(x=$(Ref(0))[]) # 2.036 ns
```
Calling a complex function is slower than calling a simple function, even if the complex portion is not run. Perhaps RadixSort is more complex than QuickSort in this sense.

For comparison, MergeSort allocates a temporary buffer before switching for dispatch to InsertionSort, so for size less than 20, it allocates an unnecessary buffer, but nobody seems to be complaining about this. (The allocation results in a roughy 10x slowdown)

If folks want high-speed sorting of very short arrays, SortingNetworks.jl and other techniques can be much faster than what we have in Base.

And also, adding a fallback to InsertionSort for short arrays before dispatching to `fpsort!`'s techniques solves the issue for small vectors. After 7f2b07696fe69ccfe4ab12293b06e79748ba974f we have:

```
@benchmark sort!(x) setup=(x=rand(1)) # 100% speedup
@benchmark sort!(x) setup=(x=rand(3)) # 59% speedup
@benchmark sort!(x) setup=(x=rand(10)) # 8% speedup
@benchmark sort!(x) setup=(x=rand(30)) # 7% slowdown (same as original because SMALL_THRESHOLD = 20)
```

LilithHafner committed:
move set defalg earlier in Base.jjl to set up for potetial bisection

LilithHafner committed:
put defalg back together!

LilithHafner committed:
bisection 1

Co-Authored-By: Dilum Aluthge <dilum@aluthge.com>

LilithHafner committed:
bisection 2

LilithHafner committed:
bisection 3 (final bisection if tests pass)

LilithHafner committed:
bisection 4 (final bisection if tests pass)

LilithHafner committed:
Add specialized sorting for Bool

Sorting used to be okay for booleans, but radix sort messed with the
dispatch and made it up to 6 times slower in the worst edge case. Rather
than deal with that dispatch, I added compile time dispatch to a
specialized boolean sorting routine that works on arbitrary orderings
and is twice as fast as what was previously in base.

LilithHafner committed:
Immediately dispatch to sort_int_range! for long Vector{[U]Int8}

This is mostly noticable for lengths 257-1999 because they were
previously dispatched to comparrison sort, skipping dispatch to
sort_int_range!

Also add comments.

LilithHafner committed:
fix typo, lol

LilithHafner committed:
test and fix boolean sorting

LilithHafner committed:
fix overflow performance bug that prevented edge case dispatch to sort_int_range!

LilithHafner committed:
restrict fpsort! policy on perm orderings to exclude Bool

Because a) indexing with a Bool is not valid and b) specialized Bool sorting is faster

LilithHafner committed:
fix typo

LilithHafner committed:
switch from custom Bool sort to re-use of sort_int_range! (faster)

LilithHafner committed:
rename RadixSort to AdaptiveSort

LilithHafner committed:
also rename in tests

ViralBShah commented:
@quinnj What all do you think is necessary to get this to a point where it can be merged?

LilithHafner commented:
I'm running comprehensive benchmarking on my end with SortMark.jl, and there are still a few performance bugs if you look hard enough. I'll try to get them cleared up over the weekend.

Mostly they are a result of modifications to the mechanisms underlying the dispatching policy.

LilithHafner committed:
add new boolean sorting (instead of slower dispatch to sort_int_range!)

LilithHafner committed:
rework dispatch (and remove Some from Serializable)

LilithHafner committed:
spelling and minor comments

LilithHafner committed:
fix typo and only unsign when possible

LilithHafner commented:
Things that I think need to be done before merging:

This is a nontrivial addition, and I don't think anyone has done a full code review yet, so there's that.

I couldn't manage to get it to build without defining `defalg` in sort.jl and then redefining it in `Base.jl`. This feels weird, ugly, and unnecessary. I also don't know why it has to be the way it is, only that it fails to build when the final definitions are anywhere in sort.jl (even after the `Sort` module). I would like it if someone more familiar with bootstrapping lends a hand with that.

The whole point of this is performance improvement, so we should probably do a nanosoldier run.
@nanosoldier `runbenchmarks(ALL, vs=":master")`

Things that would be nice but I don't think are blocking:

This is not perfect and if you look hard enough you can still find substantial regressions. The worst I've found so far is `@benchmark sort!(x) setup=(x=rand([typemin(UInt), typemax(UInt)], 3000))` with a 2x slowdown, but if we have a different distribution, this is still an improvement: `@benchmark sort!(x) setup=(x=rand(UInt, 3000))` has a 2x speedup and `@benchmark sort!(x) setup=(x=rand(0:10^5, 3000))` has a 5x speedup.

This is also not complete in that it doesn't address the performance issues of sortperm. In general, I have not made all optimizations or applied the new algorithm in all cases where it is beneficial. I expect further improvement here is possible and perhaps even a low hanging fruit. But I don't think that needs to be taken care of before merging.

As of now, I don't know of any bugs in the PR or large gaps in testing. Nevertheless, both of these may well be present. If someone else wants to poke around a bit looking for regressions or bugs, that would be great!

ViralBShah commented:
One question that came to my mind - should this go to SortingAlgorithms.jl? That seems to already have a radix sort, where this might be an improvement?

LilithHafner commented:
It comes down to how much we care about fast sorting. I think that sorting is a common enough operation that we should have fast sorting algorithms available in Base, but I'll let other people weigh in if they want to.

For comparison, we already have about 140 lines in the submodule Sort.Float dedicated to optimizing floating-point and missing+floating point sorting, and about 60 lines dedicated to quick-sort which is an optimization for unstable sorting.

Sort.Float is about a 2x speedup within its domain based on `@benchmark sort!(x, 1, 1000, Base.QuickSort, Base.Forward) setup=(x=rand(1000))` vs `@benchmark sort!(x) setup=(x=rand(1000))` and it necessitates extending the sorting internals to pass around lo and hi endpoints to accommodate the way it operates and disambiguate methods. QuickSort is also about a 2x speedup on its domain based on `@benchmark sort!(x, alg=Base.QuickSort) setup=(x=rand(Int, 1000))` vs `@benchmark sort!(x, alg=Base.MergeSort) setup=(x=rand(Int, 1000))`. On RadixSort's domain (large arrays or narrow bitwidth), it can be a x2 speedup or much more.

oscardssmith commented:
@ViralBShah imo, this belongs in base, not a package. SortingAlgorithms is great for things that are niche, but a good radix sort is the best sorting algorithm for most common uses.

LilithHafner committed:
bump to rerun CI

LilithHafner committed:
undo bump to rerun CI

LilithHafner committed:
update comment to explain defalg redefinition

LilithHafner committed:
Merge branch 'JuliaLang:master' into patch-9

LilithHafner committed:
micro optimization

2.5% speedup on `@belapsed sort!(x) setup=(x=rand(Int, 2001))`

LilithHafner committed:
Style

LilithHafner committed:
new dispatch (before exterma-serialization separation

LilithHafner committed:
Seperate extrema and serialization in all cases

LilithHafner committed:
Minor simplification (maybe helps performance 1-2%)

LilithHafner committed:
Add comments to radaix_sort! and make computationally negligable changes to help the code match the explanation

radix_sort! is not exported and only called with Int lo and hi values, 
so changing the eltype of counts to UInt should have no exported impact. 
I made this change 
a) to simplify slightly
b) because unsigned fails for BitInt
c) because line 715 can now overflow as of the previous commit, and if 
we are using a custom integer type, that might error.

unsigned(hi-lo) gives a false illusion of generality.

LilithHafner committed:
Update docstrings & comments

LilithHafner committed:
finish type change: unsigned(eltype(hi-lo)) -> UInt

LilithHafner committed:
minor comment style

LilithHafner committed:
reorder for readability

LilithHafner committed:
Integrate with compiler bootstrapping & remove defalg redefinitions!

LilithHafner committed:
Remove unhelpful comment

LilithHafner committed:
remove stray newline

LilithHafner committed:
Code review based on PR diff

LilithHafner committed:
Spelling

LilithHafner committed:
Qualify DEFAULT_(UN)STABLE in tests

LilithHafner committed:
Comments and such

LilithHafner commented:
@oscardssmith, @quinnj, @ViralBShah I think this is ready for review if any of you are willing to look it over. It also needs a nanosoldier run and I believe I don't have the permissions for that.

LilithHafner commented:
Bump :)

oscardssmith commented on base/sort.jl:
```diff
@@ -652,13 +669,203 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::PartialQuickSort,
+function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::AdaptiveSort, o::Ordering)
+    # if the sorting task is unserializable, then we can't radix sort or sort_int_range!
+    # so we skip straight to the fallback algorithm which is comparison based.
+    U = Serial.Serializable(eltype(v), o)
```
this seems very weird.

LilithHafner replied:
`Sort.Serial.Serializable` is a trait that indicates both whether something is serializable and to what datatype it serializes to. For example, by default, sorting by the function `f(x) = x & 0xff` is serializable, but if someone specialized `Base.Sort.Serial.serialize` and `Base.Sort.Serial.deserialize` for `f`, then they could also specialize `Base.Sort.Serial.Serializable` to return `UInt8`.

AdaptiveSort uses this trait to determine if any of its optimizations are possible.

nalimilan replied:
I also find that "serialize" could be confusing as it's completely different from `Base.serialize`. Maybe we could find another term, which more explicitly indicates that this maps a value to an unsigned integer?

LilithHafner replied:
SortingAlgorithms.jl uses `uint_mapping`. A key facet of this implementation is the mapping is reversible, which led me to the term serialization. I'm not a fan of `UIntBijection`.

LilithHafner replied:
Perhaps `Serial` -> `UIntMapping`, `serialize` -> `uint_map`, `deserialize` -> `uint_unmap`, `Serializable` -> `UIntMappable`?

oscardssmith replied:
that sounds good

nalimilan replied:
+1

Maybe these don't need to go into a submodule BTW.

LilithHafner replied:
The functions do all have "uint_map" in their name, so there shouldn't be an issue with taking them out of a submodule. I was following the example of `Float`, but `Float` defines things like `Left` and `specials_to_end` which are once to keep in a Submodule and `Serial`/`UIntMapping` does not define this sort of thing.

LilithHafner replied:
Renamed and removed the submodule.

oscardssmith commented on base/sort.jl:
```diff
@@ -652,13 +669,203 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::PartialQuickSort,
+    # to avoid introducing excessive detection costs for the trivial sorting problem
+    # and to avoid overflow, we check for small inputs before any other runtime checks
+    hi <= lo && return v
+    ln = maybe_unsigned(hi-lo) # ln == length(v[lo:hi])-1. Lower number = fewer letters :)
```
rename `ln` to `len`

LilithHafner replied:
I'll do that that. Is it okay that `len` is a proxy for `length-1`?

oscardssmith replied:
`lenm1` maybe?

LilithHafner replied:
Ugly, but accurate. The concept is a bit ugly. I think `lenm1` is probably best (`lengthm1` is misleading because that implies `length(v)`). The reason I'm using this at all is computing `hi-lo+1` doesn't substantively change dispatch, has a (detectable for very small arrays?) cost, and, most importantly, risks overflow.

LilithHafner replied:
Following this pattern would entail changing `rln` to `rangelenm1` and `urln` to `urangelenm1` or `unsigned_range_len_m1`. I want to be consistent, descriptive, and ideally not use 4 words.

LilithHafner replied:
from statistics, `range=max-min`, so `range_len_m1` can just be `range`. The reason `sort_int_range!` uses `rangelen` is because it is allocating an array of length `rangelen`, and `rangelen` is a proxy for `range + 1`.

In conclusion
`ln` -> `lenm1`
`rln` -> `range`
and better still:
`mn -> v_min`
`mx -> v_max`
`rln -> v_range`
`umn -> u_min`
`umx -> u_max`
`urln -> u_range`

nalimilan replied:
"range" should rather be called "span", right?

oscardssmith replied:
yeah

LilithHafner replied:
The concept I'm referring to with the variable name `range` is `max-min`. I believe that this is the unambiguous statistical definition of range, while span is a more ambiguous term. For this reason, I chose to follow @JeffBezanson's choice in 2016 to use the term range for this concept. (Jeff was referring to `max-min+1` and called it `rangelen`.)

Is there some commonly accepted definition of span or range that I'm not aware of? Why do y'all prefer span?

https://en.wikipedia.org/wiki/Range
https://en.wikipedia.org/wiki/Range_(statistics)
https://en.wikipedia.org/wiki/Span

nalimilan replied:
Nevermind, I thought this was a standard term but looks like it isn't. We have a function `span` in StatsBase, but it returns... a `min:max` range. :-D

oscardssmith commented on base/sort.jl:
```diff
@@ -652,13 +669,203 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::PartialQuickSort,
+        end
+    end
+
+    umn, umx = Serial.serialize(mn, o), Serial.serialize(mx, o)
```
why are we serializing?

LilithHafner replied:
To compute `umx-umn` which informs the decision about whether to use `InsertionSort` or `sort_int_range!` or `RadixSort`. If `umx-umn` is small, then value based methods are more likely to outperform comparison based methods.

LilithHafner replied:
The variables `umn` and `umx` are the extrema of the array `u`, but the decision about whether to compute `u` at all depends on `umn` and `umx`. It's backward from a readability standpoint.

oscardssmith replied:
makes sense.

oscardssmith commented:
Overall this mostly looks good to me. The biggest thing I think it needs is a good set of benchmarks across a wide range of sizes and distributions.

LilithHafner committed:
Rename variables

ln -> lenm1
mn -> v_min
mx -> v_max
rln -> v_range
umn -> u_min
umx -> u_max
urln -> u_range

LilithHafner commented:
For development, I've been using SortMark.jl which takes some shortcuts to allow testing a large cross product of inputs. It has served me well, but for validation, I think it makes sense to use established techniques (e.g. BencharkTools & Nanosoldier).

I'm thinking sizes `10 .^ (0:.5:7)`
Distributions (inspired by SortPerf.jl): 
- random
- sorted
- reverse sorted
- all same
- 4 unique values
- 4 unique values close to each other (`1:len÷3`)
- short int range (`1:len`)
- very sort int range (`1:len÷3`)
- somewhat sort int range (`1:3len`)
- sorted with 3 random exchanges
- sorted with 10 random elements appended
- pathological for quicksort

benchmarked with `@benchmark sort!(x) setup=$f($len)) evals=1`

Maybe It should also span `Int`, `Float64`, `Float32`, `UInt8`, `Int128`, and perhaps other types.
And maybe it should span multiple orderings (e.g. Forward, Reverse, By, Perm)

Too many axes is a computational problem, though.

LilithHafner committed:
Find, test & fix bug

LilithHafner commented:
Unfortunately, benchmarking is not super straightforward:
```
  1 "sort(x) setup=(x=rand(Int, 1000)) evals=100 samples=1000" => TrialJudgement(+29.30% => regression)
  2 "sort!(copyto!(y, x)) setup=(x=rand(Int, 1000); y=similar(x)) evals=100 samples=1000" => TrialJudgement(+38.92% => regression)
  3 "sort!(x) setup=(x=rand(Int, 1000)) evals=1" => TrialJudgement(-39.67% => improvement)
  4 "sort!(shuffle!(x)) setup=(x=rand(Int, 1000)) evals=100 samples=1000" => TrialJudgement(-31.68% => improvement)
  5 "sort!(x) setup=(x=rand(Int, 1000)) evals=100 samples=1000" => TrialJudgement(+427.58% => regression)
  ```
1, 2, 3, and 4 all seem like reasonable choices to me, but they have substantially different results. (A 39% improvement is 1.5x as big as a 39% regression from a logarithmic perspective)

nalimilan commented on base/sort.jl:
```diff
@@ -426,6 +426,23 @@ struct InsertionSortAlg <: Algorithm end
+  * Insertion sort for short lists
+  * Radix sort for long lists
+  * Counting sort for lists of integers spanning a short range
+  * Fallback algorithm when the input cannot be efficiently
+    converted into a list of integers maintaining sort order
```
I don't understand this. Does this refer to the `fallback`? Also "list" isn't typical Julia terminology, better say "vector" (here and below).

LilithHafner replied:
I've rewritten it. Does the docstring make sense now?

nalimilan commented on base/sort.jl:
```diff
@@ -652,13 +669,203 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::PartialQuickSort,
+#
+# In the case of an odd number of passes, the returned list will === the input list t,
+# not v. This is one of the many reasons radix_sort! is not exported.
+function radix_sort!(v::AbstractVector{U}, lo::Integer, hi::Integer, bits::Unsigned,
+               ::Val{CHUNK_SIZE}, t::AbstractVector{U}) where {U <: Unsigned, CHUNK_SIZE}
```
```suggestion
function radix_sort!(v::AbstractVector{U}, lo::Integer, hi::Integer, bits::Unsigned,
                     ::Val{CHUNK_SIZE}, t::AbstractVector{U}) where {U <: Unsigned, CHUNK_SIZE}
```

nalimilan commented on base/sort.jl:
```diff
@@ -652,13 +669,203 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::PartialQuickSort,
+        # of bucket 0x00 in t[counts[1]], the next element of bucket 0x00 in t[counts[1]+1],
+        # and the last element of bucket 0x00 in t[counts[2]-1].
+
+        for k in lo:hi # Is this iteration slower than it could be?
```
What does this question mean? Either make it more explicit or drop it?

LilithHafner replied:
When profiling I found that the actual iteration step seems to be taking longer than it should (perhaps because it is the only branch). I tried hacking in a [duff's device ](https://en.wikipedia.org/wiki/Duff%27s_device) but (as expected) that didn't change anything. At this point, I'm going to answer the question with "seems pretty decent to me" and take it out of the source code.

LilithHafner replied:
Done

nalimilan commented on base/sort.jl:
```diff
@@ -652,13 +669,203 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::PartialQuickSort,
+    v
+end
+
+function sort!(v::AbstractVector{<:Bool}, lo::Integer, hi::Integer, a::AdaptiveSort, o::Ordering)
```
This is counting sort, right? Add a comment to mention this?

LilithHafner replied:
done.

nalimilan commented on base/sort.jl:
```diff
@@ -652,13 +669,203 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::PartialQuickSort,
+maybe_unsigned(x::Union{Int8, Int16, Int32, Int64, Int128}) = unsigned(x)
+function _extrema(v::AbstractArray, lo::Integer, hi::Integer, o::Ordering)
+    mn = mx = v[lo]
+    while lo < hi
+        lo += 1
+        vi = v[lo]
```
I guess `@inbounds` is OK here?
```suggestion
    @inbounds for i in lo:(hi-1)
        vi = v[i]
```

LilithHafner replied:
`@inbounds` is okay, yes. I couldn't find any performance improvements from using it so I didn't. Looking deeper, at your prompting, I did find some. According to my benchmarks the following three versions have a runtime of 1ns per element for `rand(10_000)` under the ordering `Base.Sort.Float.Left()`:
```julia
while lo < hi
        lo += 1
        vi = v[lo]
```
```
@inbounds while lo < hi
        lo += 1
        vi = v[lo]
```
```
@inbounds for i in lo+1:hi
        vi = v[i]
```
and this one has a runtime of 1.2ns per element:
```
for i in lo+1:hi
        vi = v[i]
```
When computing the `_extrema` of 10_000 Int64s, those with the `@inbounds` annotation ran at 0.20ns, and those without ran at 0.23ns. Based on this, I accept your recommendation (with indexing fix) as an improvement in readability and speed. (aside: 0.20ns is the fastest benchmarking result I have ever seen. The `@code_llvm` instructions are pretty complex.)
```suggestion
    @inbounds for i in (lo+1):hi
        vi = v[i]
```

oscardssmith replied:
less than 1ns means the compiler broke your benchmark

LilithHafner replied:
My methodology is
```julia
(@belapsed Base.Sort._extrema(x,1,10_000,Base.Forward) setup=(x=rand(Int,10_000)))/10_000
# 2.0268e-10
```
My conclusion is that the runtime _per element_ is `.20ns`. I believe the compiler optimized effectively but did not break the benchmark.

LilithHafner replied:
Strangely, my implementation seems to be slightly faster than `extrema`. Perhaps this has something to do with the shortcuts I took assuming that the input was non-empty?
Methodology:
```julia
@belapsed extrema(x) setup=(x=rand(Int,10_000))
# 2.4671111111111115e-6
@belapsed Base.Sort._extrema(x, 1, 10_000, Base.Forward) setup=(x=rand(Int,10_000))
# 2.0268e-6
```

LilithHafner replied:
I moved this conversation to https://github.com/JuliaLang/julia/issues/44606

nalimilan commented on base/sort.jl:
```diff
@@ -711,31 +918,19 @@ function sort!(v::AbstractVector;
-        n = length(v)
-        if n > 1
-            min, max = extrema(v)
-            (diff, o1) = sub_with_overflow(max, min)
-            (rangelen, o2) = add_with_overflow(diff, oneunit(diff))
```
Why don't you use these functions when computing similar quantities in your new code?

LilithHafner replied:
See https://github.com/JuliaLang/julia/pull/44230/#discussion_r825364777

Note that the method I use dispatches to count sort on `rand(Int8, 1000)`, but the old method overflows and so fails to dispatch to counting sort.

nalimilan commented on base/sort.jl:
```diff
@@ -711,31 +918,19 @@ function sort!(v::AbstractVector;
 
 # sort! for vectors of few unique integers
-function sort_int_range!(x::AbstractVector{<:Integer}, rangelen, minval, maybereverse)
+function sort_int_range!(x::AbstractVector{<:Integer}, rangelen, minval, maybereverse, lo=firstindex(x), hi=lastindex(x))
```
```suggestion
function sort_int_range!(x::AbstractVector{<:Integer}, rangelen, minval, maybereverse,
                         lo=firstindex(x), hi=lastindex(x))
```

nalimilan commented on base/sort.jl:
```diff
@@ -1109,12 +1304,114 @@ function sort!(A::AbstractArray;
+Return `typeof(serialize(x::T, order))` if [`serialize`](@ref) and
+[`deserialize`](@ref) are implemented.
+
+If either is not implemented, return nothing.
```
```suggestion
If either is not implemented, return `nothing`.
```

nalimilan commented on base/sort.jl:
```diff
@@ -1109,12 +1304,114 @@ function sort!(A::AbstractArray;
+# that there are no NaN values
+
+# Booleans
+# serialize! fails on AbstractVector{Bool} because the reinterpret is sketchy. try:
+#     u = reinterpret(UInt8, trues(10)); u[3] = 0x17; print(u)
```
What's the problem exactly with `0x17` being converted silently to `0x01`? Isn't it enough to map `false` to `0x00` and `true` to `0x01`?

LilithHafner replied:
`sort!(collect(trues(501)), alg=Sort.AdaptiveSort(QuickSort), rev=true)` used to throw an error because `false` went to `0xff` and `true` went to `0xfe`. Apparently, that's not an issue anymore. We still never use serialize on Bool in Base because we count sort instead. We should either delete Bool entirely without comment (bc it isn't used) or leave it in (for completeness). I lean toward the former.

nalimilan commented on base/sort.jl:
```diff
@@ -1109,12 +1304,114 @@ function sort!(A::AbstractArray;
+    u
+end
+
+function deserialize!(v::AbstractVector, u::AbstractVector{U},
+    lo::Integer, hi::Integer, order::Ordering, offset::U=zero(U)) where U <: Unsigned
```
```suggestion
function deserialize!(v::AbstractVector, u::AbstractVector{U},
                      lo::Integer, hi::Integer, order::Ordering, offset::U=zero(U)) where U <: Unsigned
```

nalimilan commented on base/sort.jl:
```diff
@@ -1140,6 +1437,18 @@ right(o::Perm) = Perm(right(o.order), o.data)
 lt(::Left, x::T, y::T) where {T<:Floats} = slt_int(y, x)
 lt(::Right, x::T, y::T) where {T<:Floats} = slt_int(x, y)
 
+Serial.serialize(x::Float32, ::Left) = Serial.serialize(reinterpret(Int32, x), Reverse)
```
Why `Int32` rather than `UInt32`? Maybe worth a comment?

LilithHafner replied:
Why is so that elements serialized with Left would compare properly with elements serialized with Right. This is unnecessary, so I've simplified it. It now serializes directly and efficiently to a UInt.

nalimilan commented on base/sort.jl:
```diff
@@ -1221,6 +1530,8 @@ issignleft(o::ReverseOrdering, x::Floats) = lt(o, x, -zero(x))
 issignleft(o::Perm, i::Integer) = issignleft(o.order, o.data[i])
 
 function fpsort!(v::AbstractVector, a::Algorithm, o::Ordering)
+    length(v) < 10 && return sort!(v, first(axes(v,1)), last(axes(v,1)), SMALL_ALGORITHM, o)
```
Add a comment about performance? Why wasn't this done before (seems not directly related to radix sort)?

LilithHafner replied:
It's not directly related to radix, but at some point, radix introduced a bit of overhead in dispatch and this offset a regression. I'll add a comment abt performance.

nalimilan commented on base/sort.jl:
```diff
@@ -426,6 +426,23 @@ struct InsertionSortAlg <: Algorithm end
 struct QuickSortAlg     <: Algorithm end
 struct MergeSortAlg     <: Algorithm end
```
Would it make sense/be possible to add a `RadixSortAlg` type? Maybe it's not super useful as it's used by default, but it would be good for consistency and e.g. for benchmarking (to ensure the intended algorithm is used). Of course it would have to throw an error for types that are not supported.

LilithHafner replied:
There are simple common reasons for folks to manually dispatch to `QuickSort` instead of `MergeSort` (if they value performance over stability), but I see no such use case for manual choice of radix sort. 

I was following the patter of counting sort (`sort_int_range!`) which, like radix sort, is dynamically dispatched to, requires some precomputed values, and cannot be user specified by the `alg` keyword.

Right now, the necessary preprocessing is interleaved with the decisions about which algorithm to dispatch to. It's possible to add a `RadixSortAlg` which forces those dispatch decisions but this is much harder than it is for `QuickSort` and `MergeSort`.

If we ever want to remove radix sort from base (e.g. a better non-comparison based algorithm comes up), it will be nice to never have defined a user-facing `RadixSorAlgt` type. 

`RadixSortAlg` would be convenient to have for benchmarking, as would `CountSortAlg`, but I think those methods belong in `contrib/` or `SortPerf.jl`, not `base/sort.jl`.

LilithHafner replied:
Tentatively, I think we should move the radix sort chunk size heuristic and switch based dispatch out of `sort!(... ::AdaptiveSort ...)` so that other folks can use it in their benchmarking definitions of `sort!(... ::RadixSortAlg ...)`, but not define `RadixSortAlg` in base.

LilithHafner replied:
Done. The `chucnk_size` argument to `radix_sort!` is now optional.

nalimilan replied:
OK. Let's hear what others think.

LilithHafner replied:
@oscardssmith, what do you think about adding a `RadixSortAlg` type?

ViralBShah replied:
What did we decide here? Close this review?

oscardssmith replied:
I think it might be useful to have `NonCompareSort` or something similar that doesn't tie us to using `RadixSort` forever, but I also don't really care as long as the thing we give to users by default is fast.

nalimilan reviewed (commented):
Impressive work! I agree we need radix sort in Base as it's often the most efficient algorithm, which means we are currently slower than e.g. R for basic types.

Here are mostly formal remarks. I cannot really comment on correctness but testing is the best approach anyway. It's worth using different strategies for correctness tests (which should cover all corner cases but do not need to check all possible sizes as long as all code paths are covered), and performance tests (which do not need to cover all cases even if they should be reasonably complete).

Finally, it would be interesting to handle `Union{T, Missing}` vectors by using a first pass to put `missing` at the end like `fpsort` currently does. Have you thought about this? It can go into another PR of course. Likewise, using radix sort with `sortperm` would be nice e.g. to sort data frames on a column.

LilithHafner committed:
Update base/sort.jl

Co-authored-by: Milan Bouchet-Valat <nalimilan@club.fr>

LilithHafner commented on base/sort.jl:
```diff
@@ -652,13 +669,203 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::PartialQuickSort,
+    v_min, v_max = _extrema(v, lo, hi, o)
+    if eltype(v) <: Integer && o isa DirectOrdering
+        R = o === Reverse
+        v_range = maybe_unsigned(R ? v_min-v_max : v_max-v_min)
```
Because it is possible that the range/span overflows in the check for dispatch to count sort which would result in a critical dispatch error (time & space complexity > `typemax(Int)`). Once that dispatch branch is triggered, we know that the range/span is less than the length of the input vector. Assuming that the input vector has length less than `typemax(Int)`, it's safe to convert this to `Int`. If not, the conversion throws an error. I'm pretty sure this assumption is implicit everywhere, but I can't test it because I don't have a 32-bit machine that can handle a 4GB vector, nor a 64-bit machine that can handle an 18EB vector.

Concretely, without this unsigned, we would dispatch to count sort on `rand(Int, 300)` because `v_max-v_min < 0 < div(lenm1, 2)`

nalimilan replied:
Ah of course. FWIW, you can just run the 32-bit Julia binaries on a 64-bit machine to check that 4GB vectors are handled correctly.

LilithHafner replied:
Vectors that large are not handled correctly.
`Vector{Int}(undef, 2^29)` or larger gives an invalid array size error.
`BitVector(undef, Int64(2)^32)` or larger gives an inexact error truncating to `Int32`.
Given these results, I think it's fair to assume the return type of `length` is an `Int` and throw an inexact error if it is larger than `typemax(Int)`.

LilithHafner commented:
> Finally, it would be interesting to handle Union{T, Missing} vectors by using a first pass to put missing at the end like fpsort currently does. Have you thought about this? It can go into another PR of course. Likewise, using radix sort with sortperm would be nice e.g. to sort data frames on a column.

Those are all good ideas. I think unions with missing are handled in a package somewhere, and it would make sense to look at how they do it. Supporting sortperm is definitely a good next step. Once we have that, and we decouple missing detection from fpsort!, then radix serialization is a strict superset of fpsort!'s domain, so we might be able to replace it entirely.

I'd like to put all that in (a) subsequent PR(s).

LilithHafner committed:
Update base/sort.jl

Co-authored-by: Milan Bouchet-Valat <nalimilan@club.fr>

LilithHafner committed:
Update base/sort.jl

Co-authored-by: Milan Bouchet-Valat <nalimilan@club.fr>

LilithHafner committed:
add explain small fallback in fpsort!

LilithHafner committed:
add @inbounds to _extrema

Co-authored-by: Milan Bouchet-Valat <nalimilan@club.fr>

LilithHafner commented on base/sort.jl:
```diff
@@ -652,13 +669,203 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::PartialQuickSort,
+    # Float32[2.012, 400.0, 12.345] serializes to UInt32[0x3fff3b63, 0x3c37ffff, 0x414570a4]
+    # which is reduced to UInt32[0x03c73b64, 0x00000000, 0x050d70a5] using only 26 bits.
+    # the overhead for this subtraction is small enough that it is worthwhile in many cases.
+    @inbounds for i in lo:hi u[i] -= u_min end # this line is faster than u[lo:hi] .-= u_min
```
It is surprising.
```julia
u = rand(UInt, 10_000); lo = 1; hi = 10_000; u_min = 1729;
fast(u, lo, hi, u_min) = (@inbounds for i in lo:hi u[i] -= u_min end; u)
slow1(u, lo, hi, u_min) = (u[lo:hi] .-= u_min; u)
slow2(u, lo, hi, u_min) = (@inbounds u[lo:hi] .-= u_min; u)

@belapsed fast($u, $lo, $hi, $u_min) # 1.1795e-6
@belapsed slow1($u, $lo, $hi, $u_min) # 8.02e-6
@belapsed slow2($u, $lo, $hi, $u_min) # 7.82e-6
```

LilithHafner replied:
I put it in one line as if it were `u[lo:hi] .-= u_min`, but you're right. This isn't clean broadcasting syntax and it does require more space.
```suggestion
    # this is faster than u[lo:hi] .-= u_min
    @inbounds for i in lo:hi
        u[i] -= u_min
    end
```
I accept your suggestion with a whitespace fix.

LilithHafner committed:
multiline -= u_min

Co-authored-by: Milan Bouchet-Valat <nalimilan@club.fr>

LilithHafner commented on base/sort.jl:
```diff
@@ -652,13 +669,203 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::PartialQuickSort,
+    # Chunk sizes less than 3 should never occur, and chunk sizes greater than 8
+    # only occur for arrays of length greater than 950, and tend to occur only for arrays
+    # of length greater than about 4000 where a single dynamic dispatch is less costly
+    u2 = if chunk_size == 3
+        radix_sort!(u, lo, hi, bits, Val(0x3), t)
+    elseif chunk_size == 4
+        radix_sort!(u, lo, hi, bits, Val(0x4), t)
+    elseif chunk_size == 5
+        radix_sort!(u, lo, hi, bits, Val(0x5), t)
+    elseif chunk_size == 6 # 9% to 15% savings over dynamic dispatch
+        radix_sort!(u, lo, hi, bits, Val(0x6), t)
+    elseif chunk_size == 7 # 2% to 7% savings over dynamic dispatch
+        radix_sort!(u, lo, hi, bits, Val(0x7), t)
+    elseif chunk_size == 8 # -1% to 10% savings and common for lengths between 300 and 3000
+        radix_sort!(u, lo, hi, bits, Val(0x8), t)
```
Unfortunately, the compiler optimizes
```jl
u2 = if chunk_size <= 8
    radix_sort!(u, lo, hi, bits, Val(UInt8(chunk_size)), t)
else
    radix_sort!(u, lo, hi, bits, Val(chunk_size), t) # dynamic dispatch
end
```
to
```jl
u2 = radix_sort!(u, lo, hi, bits, Val(chunk_size), t) # dynamic dispatch
```
and triggers dynamic dispatch even for small chunk sizes. Perhaps we should simply use `u2 = radix_sort!(u, lo, hi, bits, Val(chunk_size), t) # dynamic dispatch`. It sometimes comes at a performance penalty, but it's simpler. I lean toward keeping the existing if-else chain.

LilithHafner replied:
I couldn't verify consistent speedups as a result of the compilation tracks with `Val`, so I removed them, obviating this if-else chain.

LilithHafner replied:
I'm really happy about this. The compiler hacks felt sketchy to me. I'm glad I can trust the compiler to be smarter than me about this kind of optimization.

LilithHafner committed:
add

LilithHafner commented on base/sort.jl:
```diff
@@ -652,13 +669,203 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::PartialQuickSort,
+
+    t = similar(u)
+    # This if else chain is to avoid dynamic dispatch for small cases.
+    # Chunk sizes less than 3 should never occur, and chunk sizes greater than 8
```
Good idea. Done. Perhaps it will even help the compiler follow along with the math I did justifying it.

LilithHafner committed:
Apply suggestions from code review

Co-authored-by: Milan Bouchet-Valat <nalimilan@club.fr>

LilithHafner committed:
replace list with vector in comments

LilithHafner committed:
remove unneccesary comment

LilithHafner committed:
Comment on count sort for Bool and add `@inbounds`

`@inbounds` is a 1.8x speedup for sorting trues(10_000)

LilithHafner committed:
remove unused bool serialization

LilithHafner commented on base/sort.jl:
```diff
@@ -652,13 +669,203 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::PartialQuickSort,
+    U = Serial.Serializable(eltype(v), o)
+    U === nothing && return sort!(v, lo, hi, a.fallback, o)
+
+    # to avoid introducing excessive detection costs for the trivial sorting problem
+    # and to avoid overflow, we check for small inputs before any other runtime checks
+    hi <= lo && return v
+    lenm1 = maybe_unsigned(hi-lo) # adding 1 would risk overflow
+    # only count sort on a short range can compete with insertion sort fo lenm1 < 30
+    # and the optimization is not worth the detection cost, so we use insertion sort.
+    lenm1 < 30 && return sort!(v, lo, hi, SMALL_ALGORITHM, o)
```
That makes sense. I tried inserting
```julia
    @noinline _sort!(v, lo, hi, a, o, U, lenm1)
end
function _sort!(v, lo, hi, a, o, U, lenm1)
```
and found the following results for `@belapsed sort!(x) setup=(x=rand(Int, $n)) evals=1`
```
n  | result
--------
5  | +1.5%
15 | -4%
30 | -6%
31 | +3%
```
and the following results for `@belapsed sort!(x) setup=(x=rand(31)) evals=1`
```
n  | result
--------
5  | -2.5%
30 | -3.5%
31 | -20% to +19% depending on the time of day or some other factor I don't understand
```
where negative is an improvement. In the past, I've run into spurious benchmarking results on the order of 5% with my benchmarking practices on my computer, so I'd conclude that the function break does not help substantially, and may not help at all. For comparison, the direct dispatch to insertion sort for fewer than 30 elements is a 1.7x speedup over julia-1.8's use of QuickSort dispatching to countsort for `@belapsed sort!(x) setup=(x=rand(Int, 5)) evals=1`.

I do not believe this is a worthwhile optimization.

LilithHafner committed:
bugfix: add using @assert

LilithHafner committed:
simplify float serailization

LilithHafner committed:
Remove compilation tricks and put upper limit on chunk_size

I couldn't verify performance gains from making CHUNK_SIZE and MASK known at compile time, so I'm trying to resist the urge to abuse the compiler.
I also noticed this was 20% slower than SortingAlgorithms.jl's RadixSort for `@belapsed sort!(x) setup=(x=rand(Int, 3000000)) evals=1` because it used a chunk size of 13 which is too high. On my computer the best chunk size for that case is 10, and I couldn't find a size where higher than 10 was better than 10. Once I set max chunk size to 10, the 20% regression turned into a 2.5% regression, well within margin of error.

LilithHafner commented:
Preliminary benchmarking results!
[benchmarks](https://gist.github.com/LilithHafner/d1ea8b4630c8ad2c8e72884b0bb18232)
[results](https://gist.github.com/LilithHafner/d1ea8b4630c8ad2c8e72884b0bb18232)
I was getting hard-to-reproduce segfaults using BenchmarkTools's advanced functionality, so I settled on something primitive.

Takeaways for Int64: 
for unstructured data, radix sort is typically 2x faster, sometimes up to 10x, and for pathological, >100x faster
***For structured data, especially presorted and almost presorted, radix sort is a 2x–6x regression.***

A hack that would fix these benchmarks would be to add an order detection pass that counts how many pairs of sequential elements are out of order. For example:
```julia
out_of_order = count(lt.(o, v[lo+1:hi], v[lo:hi-1])
out_of_order == 0 && return v
out_of_order == range-1 && return _reverse(v, lo, hi)
out_of_order < 4log2(range) && return sort!(v, lo, hi, InsertionSort, o)
out_of_order > range-3log2(range) && return _reverse(sort!(v, lo, hi, InsertionSort, Reverse(o)), lo, hi)
if out_of_order < range/4 || out_of_order > range*3/4
    return sort!(v, lo, hi, a.fallback, o)
end
```
I'm worried this hack might fool benchmarks but fail in real world situations.

nalimilan commented on base/sort.jl:
```diff
@@ -1240,7 +1535,7 @@ fpsort!(v::AbstractVector, a::Sort.PartialQuickSort, o::Ordering) =
 sort!(v::FPSortable, a::Algorithm, o::DirectOrdering) =
     fpsort!(v, a, o)
-sort!(v::AbstractVector{<:Integer}, a::Algorithm, o::Perm{<:DirectOrdering,<:FPSortable}) =
+sort!(v::AbstractVector{<:Union{Signed, Unsigned}}, a::Algorithm, o::Perm{<:DirectOrdering,<:FPSortable}) =
```
Why change this? Even if `Bool` has a faster method, no need to disallow it as the most specific method will take precedence.

LilithHafner replied:
Because the method actually doesn't work for Bool. This errors on master: `sort!(BitVector(undef, 10), order=Base.Perm(Base.Forward, [4.5, 2.0]))`.

LilithHafner replied:
`ERROR: ArgumentError: invalid index: false of type Bool`

nalimilan commented:
> Preliminary benchmarking results!
> [benchmarks](https://gist.github.com/LilithHafner/d1ea8b4630c8ad2c8e72884b0bb18232)
> [results](https://gist.github.com/LilithHafner/d1ea8b4630c8ad2c8e72884b0bb18232)

Interesting! Though the two links are the same.

> I was getting hard-to-reproduce segfaults using BenchmarkTools's advanced functionality, so I settled on something primitive.

Segfaults are kind of worrying. Have you checked with `--check-bounds=yes` that they don't come from bugs in your code?

Regarding partially pre-sorted inputs, I feel like having some heuristic to fall back to another algorithm is worth it if the overhead isn't too large, as it's not completely unrealistic to have mostly sorted data. Do you know how costly the check is? Why do you think it would fail in real life? For fully pre-sorted inputs, we already check `issorted` before doing anything, right?

There also seem to be limited regressions when there are unique 3 unique values. Though it's probably acceptable.

LilithHafner commented:
([benchmarks](https://gist.github.com/LilithHafner/12de62677c205218f7f1fcf30ebf3233) link editted.

I only ever got segfaults on master, not on this PR, I highly doubt this PR fixes them, but because they are resent on master, I doubt this PR introdues them either.

We never check `issorted`, though InsertionSort runs for only 1.7x as lon  `issorted` on sorted inputs. QuickSort also runs fast on sorted input (1.5x-2.5x as long as InsertionSort) because these lines let it partition quickly when data is already partitioned or almost partitioned:
```julia
        while lt(o, v[i], pivot); i += 1; end;
        while lt(o, pivot, v[j]); j -= 1; end;
```
and because its base case fallback is InseritonSort.

Yes, I was lumping 4-unique in with "structured data" and didn't mention it specifically because it is wasn't as large of a regression. Two pivots in quicksort can get 4-unique data in order, at which point the rest is pretty quick. I'm also not too concerned about those regressions. The regresssions are minor, and the domain of the regressions is confined to 4 unique values spread far apart within the bit-value-space with lengths 1e2–1e6. I ran with 20-unique and found a maximum regression of 18% at 316 elements, quickly falling off to less than 10% for len <= 100 or 1000 <= len and an improvement for len <= 32 and 10_000 <= len.

LilithHafner committed:
ckeck for ordered inputs

LilithHafner committed:
use view not macro view

LilithHafner commented:
> if the overhead isn't too large

This check slips into the custom `_extrema` function, adding an additional 40-100% to the runtime of _extrema, which divides out to about 2-4% of the runtime of sorting. Head to head benchmarking of [17f45e8](https://github.com/JuliaLang/julia/pull/44230/commits/17f45e8f02e2e93927627655ed84b2d138f98f83) vs [4931881](https://github.com/JuliaLang/julia/pull/44230/commits/49318813923c9432d7a49fc839e229189857df2b) (before vs. after the order check) fails to reveal a consistent difference except where fallbacks are triggered. This is consistent with a 2-4% cost which is undetectable with my benchmarking methods.

LilithHafner commented:
Updated [benchmarks](https://gist.github.com/LilithHafner/12de62677c205218f7f1fcf30ebf3233) and [results](https://gist.github.com/LilithHafner/d1ea8b4630c8ad2c8e72884b0bb18232).

Much better now for presorted, reverse sorted, all same, and almost sorted, but there is a new noticeable overhead in some cases.

I reran the two worst regressions and found them not nearly as bad as reported (I am benchmarking unshielded on a noisy system):
```
Int64, four_unique_close, 100000000 : 1.353740641 / 0.395264016 = 3.425   REGRESSION
this time -> 1.8x (regression)
Int64, sorted_with_10_random, 10000000 : 0.349346034 / 0.153613341 = 2.274   REGRESSION
this time -> .86x (improvement)
```
methodology:
```
path/to/julia -e 'using BenchmarkTools; four_unique_close(T, len) = rand(rand(T(1):T(max(1,len÷3)), 4), len); display(@benchmark sort!(x) setup=(x=four_unique_close(Int, 10^8)) evals=1 seconds=10)'
```
and
```
path/to/julia -e 'using BenchmarkTools; function sorted_with_10_random(T, len)
    v = rand(T, len)
    sort!(@view v[1:max(1, len-10)])
    v
end; display(@benchmark sort!(x) setup=(x=sorted_with_10_random(Int, 10^7)) evals=1 seconds=10)'```

LilithHafner committed:
decouple _extrema and out_of_order to eliminate overhead (and optimization) in cases where the existing methods are already very fast (i.e. short int range and short vectors)

LilithHafner committed:
put back all same check (free)

LilithHafner committed:
Update heuristic

notes:
source len ratio (>1 is regression)
S = sort R = reverse a = almost
float  40   .87
Int    40   .96
1:40   40  1.07
1:20   40  1.7
1:3    40  2.61
1:2    40  1.92
1:1    40  1.72
(1.734)
S Int  40   .92
R Int  40   .92
aS Int 40   .84
aR Int 40   .93
(.9027)
aR flo 40   .89
aS flo 40   .73
S floa 40   .77
R floa 40   .81
(.7979)

LilithHafner committed:
Move second dispatch to insertion sort to after presorted detection

to address regressions for sorting reverse sorted or almost reverse 
sorted data. Detection cost 2-7%, only impacting sorting vectors longer 
than 40 and shorter than 3bits (e.g. `rand(Int, 100)`).

Also put in all same detection for 128 bit types under direct order 
(free).

LilithHafner committed:
Change length threshold from len > 61 to len > 60 for consistency

LilithHafner commented:
Same links, updated content: [benchmarks](https://gist.github.com/LilithHafner/12de62677c205218f7f1fcf30ebf3233) and [results](https://gist.github.com/LilithHafner/d1ea8b4630c8ad2c8e72884b0bb18232).
Things look decent for Integers, but I have a feeling that floats could be faster.
I think the regression in 32 reverse sorted Int64 is acceptable. This regression is most extreme with 61 reverse sorted integers where it is a 5.6x regression. However, if sorting 14 vectors, 10 of which are random, and one each is sorted, almost sorted, reverse sorted, and almost reverse sorted, this PR has no impact (methodology: compute a weighted average of differences in ns between sorting times in ef2455076c and 3a5dc09ce4). Below 61 we do not check for sorted input because the detection cost becomes substantial.

LilithHafner committed:
Move submodule Sort.Serial to Sort and rename `serial` to `uint_map`

`Serial` -> nothing, `serialize` -> `uint_map`, `deserialize` -> 
`uint_unmap`, `Serializable` -> `UIntMappable`?

LilithHafner committed:
Docstring

LilithHafner committed:
Try to fix build error (1)

LilithHafner commented:
Here's a visual summary of the policy changes as of [dad6a26](https://github.com/JuliaLang/julia/pull/44230/commits/dad6a26d69859e91c44192f2c19ff84a21a7b3ef)
Old sorting policy:
<img width="486" alt="Old sorting policy" src="https://user-images.githubusercontent.com/60898866/158522468-0515e3c8-f496-4247-9850-c1f7b72aeca3.png">
New sorting policy for radixable types and orders:
<img width="486" alt="Sorting policy for radixable types and orders" src="https://user-images.githubusercontent.com/60898866/158522464-2a4f6417-d184-43fb-abd2-08e457eeef00.png">

LilithHafner committed:
Defer `bits` calculation till needed for readability

LilithHafner committed:
make floats UIntMappable, lol

LilithHafner commented:
Thanks! Yes. Both operations are very fast. An alternative to the sign pivot pass is to mutate each element in place via the same uint_mapping infrastructure that radix_sort uses, sort as units (radix or otherwise), and then invert the mapping. I still don't know when, if ever, uint_mapping is faster than pivoting.

Moving NaN and Missing to the end is still necessary either way.

StefanKarpinski commented:
Right, it's a different solution to the same problem. Pivoting seems likely to be strictly faster to me but only benchmarking will say definitively.

LilithHafner committed:
Test Coverage

LilithHafner committed:
Explain floating point exceptions

LilithHafner committed:
Fix argument order in docstrings

LilithHafner committed:
Fix argument order in docstrings part 2

LilithHafner committed:
Merge branch 'JuliaLang:master' into patch-9

LilithHafner committed:
Fix typo in tests

LilithHafner commented:
Updated benchmark [results](https://gist.github.com/LilithHafner/d1ea8b4630c8ad2c8e72884b0bb18232)

LilithHafner committed:
Replace partially presorted check with fully presorted check

LilithHafner commented:
I've expanded on SortPerf's almost-sorted distributions to find that QuickSort _does not_ consistently take advantage of partly presorted data.

Runtime in `ms` for length `10^5` and element type `Float64`:
distribution | RadixSort | QuickSort
---|---|--
Sorted with 10 random appended* | 2.05 | 1.0
Sorted with 10 random prepended | 2.04 | 70
Reverse sorted with 10 random appended | 2.03 | 12.5
Reverse sorted with 10 random prepended | 2.03 | 9.2
Sorted with 3 exchanges* | 2.03 | 0.66
Sorted with 20 exchanges | 2.03 | 1.03
Reverse sorted with 3 exchanges | 1.97 | 0.60
Reverse sorted with 20 exchanges | 2.02 | 0.59
Sorted* | 2.03 | 0.65
Reverse sorted* | 2.03 | 0.60
Random* | 2.04 | 5.76

*Present in past benchmarking

In light of this, I have replaced the partially presorted branch with a much simpler (and lower overhead)
```
# For most long arrays, a presorted check is essentially free (overhead < 1%)
lenm1 >= 200 && issorted(view(v, lo:hi), o) && return v
```
I picked 200 because that is about when the cost drops below 1%. I had to measure this cost by separately benchmarking `issorted` and `sort!` and dividing because my benchmarking accuracy is on the order of 5%.

I can also add
```
lenm1 >= 500 && issorted(view(v, lo:hi), ReverseOrdering(o)) && (reverse!(view(v, lo:hi)); return v)
```
if folks think it is worth the extra code complexity + compile time.

This will introduce regressions for certain orderings that QuickSort is good at.

LilithHafner commented:
Updated [benchmarks](https://gist.github.com/LilithHafner/12de62677c205218f7f1fcf30ebf3233) and [results](https://gist.github.com/LilithHafner/d1ea8b4630c8ad2c8e72884b0bb18232).

LilithHafner commented:
Bump :)

nalimilan commented on base/sort.jl:
```diff
@@ -1109,15 +1294,104 @@ function sort!(A::AbstractArray;
 end
 
+
+## uint mapping to alow radix sorting primitives other than UInts ##
```
```suggestion
## uint mapping to allow radix sorting primitives other than UInts ##
```

nalimilan commented on base/sort.jl:
```diff
@@ -1109,15 +1294,104 @@ function sort!(A::AbstractArray;
+end
+
+function uint_unmap!(v::AbstractVector, u::AbstractVector{U}, lo::Integer, hi::Integer,
+                      order::Ordering, offset::U=zero(U)) where U <: Unsigned
```
```suggestion
                     order::Ordering, offset::U=zero(U)) where U <: Unsigned
```

LilithHafner committed:
Apply suggestions from code review

Co-authored-by: Milan Bouchet-Valat <nalimilan@club.fr>

LilithHafner committed:
adjust presorted checks

LilithHafner commented:
> The advantage could be that 1) the rules are simpler, and 2) we are always fast for pre-sorted inputs, since `InsertionSort` is fast for them.

For adjusting the minimum presorted input threshold, I find your reasoning quite convincing. I like how you provide a theoretical consideration (1) and then back it up with a practical example of why this matters (2: instead of three cases with a bad middle, there are only two cases).

For reverse sorted input, this makes the algorithm marginally more complicated, though the runtime cost-benefit is huge. This also makes a nice pattern where very short (< 40) reverse-sorted input is pathological for insertion sort, medium-length (40 < len < 500) is handled decently by radix sort, and long (> 500) is handled perfectly with a reverse sorted check. Intuitively, as length increases, special-case detection improves. I added it.

LilithHafner commented:
> BTW, it would be nice to keep these benchmarks somewhere. Maybe you make a PR against BaseBenchmarks after this one is merged?

The benchmarks here are pretty similar to the benchmarks already in [BaseBenchmarks/sort](https://github.com/JuliaCI/BaseBenchmarks.jl/blob/master/src/sort/SortBenchmarks.jl). The difference is that these span more input distributions, more types, more lengths, and longer lengths, and those at BaseBenchmarks cover sorting functions other than `sort!`. How do these tests cover a wider area? they take longer to run. These tests take a few hours to run. I might contribute a sparse cross product akin to what I've done in SortMark.jl, but it's not trivial to migrate these benchmarks to BaseBenchmarks without bringing along too much bloat.

LilithHafner committed:
proof read

LilithHafner commented:
Okay, this looks good to me too. Thanks so much, @nalimilan, @oscardssmith, and everyone else who has had conversations and helped along the way!

oscardssmith commented:
thanks so much for setting this through!

nalimilan commented:
Thanks @LilithHafner! Don't hesitate if you feel like extending radix sort to `Union{T, Missing}` vectors. :-)

StefanKarpinski commented:
Brava! Excellent work here.

Thought on the issorted check: we could introduce a `sortcheck(v)` utility function that returns `1` or `0` or `-1` to indicate sorted, reverse sorted or neither. It would short-circuit, of course. I suspect that could be as fast as the `issorted` check.

LilithHafner commented:
Thanks!

I'm having trouble coming up with an implementation of `sortcheck` that is substantially faster than two `issorted` checks.

StefanKarpinski commented:
Sure enough, this slightly incorrect 8x unrolling of the inner loop is 2x faster:
```jl
function sortcheck(v::Vector, lt=isless)
    cmp::UInt8 = 0b11
    if !isempty(v)
        y = v[1]
        @inbounds for i = 2:8:length(v)
            for j = 0:7
                x = v[i+j]
                cmp &= 2*!lt(x, y) + !lt(y, x)
                y = x
            end
            iszero(cmp) && break
        end
    end
    !iszero(cmp)
end
```
If the end of the loop was handled correctly, this could work. Paying attention to memory alignment might improve the speed more. I do wish the compiler could figure that transformation out based on the fact that the inner loop has no side effects and once `cmp` becomes zero it can't come back, but I suppose that's asking a lot of LLVM.

oscardssmith commented:
The compiler problem is very similar to vectorizing `findfirst` automatically which would also be really nice.

LilithHafner commented:
I'm more concerned with runtime on entirely unsorted input than on input with long sorted prefixes. On random data, the latest sortcheck is fairly expensive:
```julia
@belapsed((sort!(x); sort(x; rev=true)), setup=(x=rand(40)))/2 #1.6455882352941177e-7
@belapsed issorted(x) setup=(x=rand(40)) #3.382e-9    2%
@belapsed sortcheck(x) setup=(x=rand(40)) #1.292778335005015e-8  8%
```

StefanKarpinski commented:
I was trying to make something faster than *two* `issorted` checks, which this is. But yes, it's still slower than just one `issorted` check, although only 6% of sort time on my system. Why are you sorting the data twice in your benchmark?

LilithHafner commented:
I'm sorting twice and dividing time by 2 because otherwise it would be repeatedly resorting sorted data, giving an unrealistically fast sorting time. On random data on my system, I'm measuring [`sortcheck`](https://github.com/JuliaLang/julia/pull/44230#issuecomment-1088867699) running 2.6x slower than two `issorted` checks.
```julia
@belapsed((sort!(x); sort(x; rev=true)), setup=(x=rand(40)))/2 #1.684660633484163e-7
@belapsed issorted(x) setup=(x=rand(40)) #3.383e-9.                 2%
@belapsed issorted(x) || issorted(x, Reverse) setup=(x=rand(40)) #4.9079079079079075e-9  3%
@belapsed sortcheck(x) setup=(x=rand(40)) #1.2937813440320962e-8    8%
@belapsed sort!(x) setup=(x=rand(40)) #1.1508637873754152e-7        70%
```
